File: frontend/src/types/research.ts
// =============================================================================
// Research Feature Types
// Mirrors backend CamelModel schemas (all camelCase)
// =============================================================================

// Workflow stage management
export type ResearchStage =
  | 'input'
  | 'searching'
  | 'results'
  | 'select-action'
  | 'proposing'
  | 'review'
  | 'applying'
  | 'done';

// =============================================================================
// Phase 1: Academic Search
// =============================================================================

export interface AcademicSearchRequest {
  query: string;
  maxResults?: number | undefined;
  academicOnly?: boolean | undefined;
  preferredTier?: number | undefined;
  unitId?: string | undefined;
}

export interface AcademicSearchResultItem {
  title: string;
  url: string;
  doi?: string | undefined;
  abstract?: string | undefined;
  authors: string[];
  publicationYear?: number | undefined;
  sourceName?: string | undefined;
  citationCount?: number | undefined;
  academicScore: number;
  provider?: string | undefined;
}

export interface AcademicSearchResponse {
  query: string;
  results: AcademicSearchResultItem[];
  totalResults: number;
  tierUsed: number;
  tierName: string;
}

export interface TierInfo {
  tier: number;
  name: string;
  available: boolean;
  reason?: string | undefined;
}

export interface AvailableTiersResponse {
  tiers: TierInfo[];
}

// =============================================================================
// Phase 2: URL Extraction
// =============================================================================

export interface UrlExtractRequest {
  urls: string[];
  summarize?: boolean | undefined;
  purpose?: string | undefined;
}

export interface ExtractedUrlItem {
  url: string;
  title?: string | undefined;
  summary?: string | undefined;
  keyPoints: string[];
  academicScore: number;
  contentType: string;
  error?: string | undefined;
}

export interface UrlExtractResponse {
  results: ExtractedUrlItem[];
  total: number;
  successful: number;
  failed: number;
}

// =============================================================================
// Phase 3: Outline Synthesis — Inputs
// =============================================================================

export interface SourceInput {
  title: string;
  url?: string | undefined;
  summary?: string | undefined;
  keyPoints?: string[] | undefined;
}

export interface ScaffoldFromSourcesRequest {
  sourceData: SourceInput[];
  unitTitle: string;
  unitDescription?: string | undefined;
  durationWeeks?: number | undefined;
  pedagogyStyle?: string | undefined;
  unitId?: string | undefined;
  designId?: string | undefined;
}

export interface CompareRequest {
  sourceData: SourceInput[];
  unitId: string;
}

export interface MatchReadingListRequest {
  sourceData: SourceInput[];
  unitId: string;
}

// =============================================================================
// Phase 3: Outline Synthesis — Scaffold Response (mirrors ai.py)
// =============================================================================

export interface ScaffoldULO {
  code: string;
  description: string;
  bloomLevel: string;
}

export interface ScaffoldWeek {
  weekNumber: number;
  topic: string;
  activities: string[];
}

export interface ScaffoldAssessment {
  title: string;
  category: string;
  weight: number;
  dueWeek?: number | undefined;
}

export interface ScaffoldUnitResponse {
  title: string;
  description: string;
  ulos: ScaffoldULO[];
  weeks: ScaffoldWeek[];
  assessments: ScaffoldAssessment[];
}

// =============================================================================
// Phase 3: Outline Synthesis — Comparison Response
// =============================================================================

export interface ComparisonWeek {
  weekNumber: number;
  topic: string;
  coverage: 'well_covered' | 'partially_covered' | 'not_covered';
  matchingSources: string[];
}

export interface ComparisonProposal {
  unitId: string;
  weeks: ComparisonWeek[];
  gaps: string[];
  overlaps: string[];
  suggestions: string[];
}

// =============================================================================
// Phase 3: Outline Synthesis — Reading List Response
// =============================================================================

export interface UnitWeekInfo {
  weekNumber: number;
  topic: string;
}

export interface ResourceMatch {
  url: string;
  title: string;
  suggestedWeek: number;
  confidence: number;
  reasoning: string;
  assignedWeek?: number | undefined;
  skipped?: boolean | undefined;
}

export interface ReadingListProposal {
  unitId: string;
  unitWeeks: UnitWeekInfo[];
  matches: ResourceMatch[];
  unmatchedCount: number;
  avgConfidence: number;
}

// =============================================================================
// Phase 3: Apply Endpoints
// =============================================================================

export interface ApplyScaffoldRequest {
  proposal: ScaffoldUnitResponse;
  unitId?: string | undefined;
}

export interface ApplyComparisonRequest {
  proposal: ComparisonProposal;
  selectedSuggestions: number[];
}

export interface ApplyReadingListRequest {
  proposal: ReadingListProposal;
  saveAsSources?: boolean | undefined;
}

export interface ApplyScaffoldResponse {
  unitId: string;
  created: boolean;
}

export interface ApplyComparisonResponse {
  unitId: string;
  appliedSuggestions: number;
}

export interface ApplyReadingListResponse {
  savedSources: number;
}

// =============================================================================
// Phase 4: Research Settings
// =============================================================================

export interface SearchApiKeys {
  googleCseApiKey?: string | undefined;
  googleCseEngineId?: string | undefined;
  braveSearchApiKey?: string | undefined;
  tavilyApiKey?: string | undefined;
  coreApiKey?: string | undefined;
}

export interface ResearchSettings {
  preferredTier: number;
  searchApiKeys?: SearchApiKeys | undefined;
  searxngUrl?: string | undefined;
}

// =============================================================================
// =============================================================================
// Phase 5: Captured Page Metadata (Embedded Browser)
// =============================================================================

export interface CapturedPageMetadata {
  url: string;
  title: string;
  description?: string | undefined;
  doi?: string | undefined;
  authors?: string[] | undefined;
  publicationDate?: string | undefined;
  publisher?: string | undefined;
  journalName?: string | undefined;
  volume?: string | undefined;
  issue?: string | undefined;
  pages?: string | undefined;
  isbn?: string | undefined;
  contentText?: string | undefined;
  academicScore?: number | undefined;
  extractionError?: string | undefined;
}

// =============================================================================
// Shared action type for SourceSelector
// =============================================================================

export type ResearchAction = 'scaffold' | 'compare' | 'reading-list';

File: backend/app/schemas/research.py
"""
Research schemas for academic search, URL extraction, and outline synthesis.
"""

from pydantic import Field

from app.schemas.ai import ScaffoldUnitResponse
from app.schemas.base import CamelModel

# =============================================================================
# Phase 1: Academic Search
# =============================================================================


class AcademicSearchRequest(CamelModel):
    """Request to search academic databases."""

    query: str = Field(..., min_length=1, max_length=1000)
    max_results: int = Field(default=20, ge=1, le=100)
    academic_only: bool = Field(default=True)
    preferred_tier: int | None = Field(
        default=None, description="Preferred search tier (1-4)"
    )
    unit_id: str | None = Field(
        default=None, description="Unit ID for context-aware search"
    )


class AcademicSearchResultItem(CamelModel):
    """A single academic search result."""

    title: str
    url: str
    doi: str | None = None
    abstract: str | None = None
    authors: list[str] = Field(default_factory=list)
    publication_year: int | None = None
    source_name: str | None = None
    citation_count: int | None = None
    academic_score: float = Field(default=0.0, ge=0.0, le=1.0)
    provider: str = ""


class AcademicSearchResponse(CamelModel):
    """Response from academic search."""

    query: str
    results: list[AcademicSearchResultItem] = Field(default_factory=list)
    total_results: int = 0
    tier_used: int = 1
    tier_name: str = "Academic"


class TierInfo(CamelModel):
    """Information about a search tier."""

    tier: int
    name: str
    available: bool
    reason: str | None = None


class AvailableTiersResponse(CamelModel):
    """List of available search tiers."""

    tiers: list[TierInfo] = Field(default_factory=list)


# =============================================================================
# Phase 2: URL Extraction
# =============================================================================


class UrlExtractRequest(CamelModel):
    """Request to extract content from URLs."""

    urls: list[str] = Field(..., min_length=1, max_length=20)
    summarize: bool = Field(default=True)
    purpose: str = Field(default="general")


class ExtractedUrlItem(CamelModel):
    """Extracted content from a single URL."""

    url: str
    title: str | None = None
    summary: str | None = None
    key_points: list[str] = Field(default_factory=list)
    academic_score: float = Field(default=0.0, ge=0.0, le=1.0)
    content_type: str = "unknown"
    error: str | None = None


class UrlExtractResponse(CamelModel):
    """Response from URL extraction."""

    results: list[ExtractedUrlItem] = Field(default_factory=list)
    total: int = 0
    successful: int = 0
    failed: int = 0


# =============================================================================
# Phase 3: Outline Synthesis (Propose/Apply)
# =============================================================================


class SourceInput(CamelModel):
    """Lightweight source data for synthesis prompts."""

    title: str
    url: str
    summary: str | None = None
    key_points: list[str] = Field(default_factory=list)


class ScaffoldFromSourcesRequest(CamelModel):
    """Request to scaffold a unit from research sources."""

    source_data: list[SourceInput] = Field(..., min_length=1)
    unit_title: str = Field(..., min_length=1, max_length=500)
    unit_description: str = Field(default="", max_length=5000)
    duration_weeks: int = Field(default=12, ge=1, le=52)
    pedagogy_style: str = Field(default="mixed_approach")
    unit_id: str | None = Field(
        default=None, description="Unit ID for Learning Design lookup"
    )
    design_id: str | None = Field(
        default=None, description="Specific Learning Design ID"
    )


class CompareRequest(CamelModel):
    """Request to compare sources against an existing unit."""

    source_data: list[SourceInput] = Field(..., min_length=1)
    unit_id: str = Field(..., description="Unit to compare against")


class ComparisonWeek(CamelModel):
    """Coverage analysis for a single week."""

    week_number: int
    topic: str
    coverage: str = Field(
        ..., description="well_covered | partially_covered | not_covered"
    )
    matching_sources: list[str] = Field(
        default_factory=list, description="Titles of matching sources"
    )


class ComparisonProposal(CamelModel):
    """AI-generated comparison of sources vs existing unit."""

    unit_id: str
    weeks: list[ComparisonWeek] = Field(default_factory=list)
    gaps: list[str] = Field(default_factory=list)
    overlaps: list[str] = Field(default_factory=list)
    suggestions: list[str] = Field(default_factory=list)


class MatchReadingListRequest(CamelModel):
    """Request to match sources to unit weeks."""

    source_data: list[SourceInput] = Field(..., min_length=1)
    unit_id: str = Field(..., description="Unit to match resources to")


class UnitWeekInfo(CamelModel):
    """Minimal week info for the reading list UI."""

    week_number: int
    topic: str


class ResourceMatch(CamelModel):
    """AI-suggested mapping of a source to a week."""

    url: str
    title: str
    suggested_week: int | None = None
    confidence: float = Field(default=0.0, ge=0.0, le=1.0)
    reasoning: str = ""
    assigned_week: int | None = Field(
        default=None, description="User-corrected week assignment"
    )
    skipped: bool = Field(default=False, description="User chose to skip this resource")


class ReadingListProposal(CamelModel):
    """AI-generated reading list mapping."""

    unit_id: str
    unit_weeks: list[UnitWeekInfo] = Field(default_factory=list)
    matches: list[ResourceMatch] = Field(default_factory=list)
    unmatched_count: int = 0
    avg_confidence: float = 0.0


class ApplyScaffoldRequest(CamelModel):
    """Request to apply a scaffold proposal."""

    proposal: ScaffoldUnitResponse
    unit_id: str | None = Field(
        default=None, description="Existing unit to overwrite, or None to create new"
    )


class ApplyComparisonRequest(CamelModel):
    """Request to apply selected comparison suggestions."""

    proposal: ComparisonProposal
    selected_suggestions: list[str] = Field(
        default_factory=list, description="Suggestion texts to adopt"
    )


class ApplyReadingListRequest(CamelModel):
    """Request to apply a reading list proposal."""

    proposal: ReadingListProposal
    save_as_sources: bool = Field(
        default=True, description="Save matched resources as ResearchSource entries"
    )


# =============================================================================
# Phase 4: Research Settings
# =============================================================================


class SearchApiKeys(CamelModel):
    """User's search API keys."""

    google_cse_api_key: str | None = None
    google_cse_engine_id: str | None = None
    serper_api_key: str | None = None
    brave_search_api_key: str | None = None
    tavily_api_key: str | None = None
    core_api_key: str | None = None


class ResearchSettings(CamelModel):
    """User's research preferences."""

    preferred_tier: int = Field(default=1, ge=1, le=4)
    search_api_keys: SearchApiKeys = Field(default_factory=SearchApiKeys)
    searxng_url: str | None = None
    excluded_domains: list[str] = Field(
        default_factory=list,
        description="Domains to drop from every tier's results (e.g. youtube.com)",
    )
